### PR TITLE
CR field for default channel

### LINKED ISSUE #23: Eventing CR field for default channel
**Problem**
 In eventing we have a configmap to update, to set/configure which channel is the default one (default-ch-webhook CM), like: https://github.com/knative/eventing/blob/master/config/core/configmaps/default-channel.yaml#L24-L31

We should better do it with a CR field, similar to `defaultBrokerClass`

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
System integrator

**Exit Criteria**

**Time Estimate (optional):**
1

**Additional context (optional)**
`defaultBrokerClass`: https://github.com/knative/eventing-operator/issues/175

### PR DESCRIPTION
**Pull request by aliok**
Fixes #23

## Proposed Changes

* New CR field for default channel template
* Defaults to `InMemoryChannel`
* Did some refactoring in brokerclass defaults code to share code with channel defaults

**Release Note**

```release-note
It is now possible to set the default channel template within KnativeEventing CR with `defaultChannelTemplate` field. Defaults to `messaging.knative.dev/v1beta InMemoryChannel`.
```

### Test instructions
Apply following:
```yaml
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  defaultChannelTemplate:
    apiVersion: messaging.knative.dev/v1alpha1
    kind: KafkaChannel
    spec:
      numPartitions: 10
      replicationFactor: 1
```
Check `default-ch-webhook` configmap.
You should see
```
data:
  default-ch-config: |
    clusterDefault:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: KafkaChannel
```

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -313,3 +314,24 @@ metadata:
 spec:
   defaultBrokerClass: MTChannelBasedBroker
 ```
+
+## spec.defaultChannelTemplate
+
+Knative Eventing allows you to define a default channel template when
+the user creates a generic `Channel`.
```
Comment (mattmoor) [inline docs/configuration.md:321]: Remove trailing whitespace:
```suggestion
the user creates a generic `Channel`.
```

File: docs/configuration.md [hunk 2]
```diff
@@ -313,3 +314,24 @@ metadata:
...
 ```
+
+## spec.defaultChannelTemplate
+
+Knative Eventing allows you to define a default channel template when
+the user creates a generic `Channel`. 
+Defaults to `messaging.knative.dev/v1beta1 InMemoryChannel` if not set.
+
+Following configuration sets default channel to `KafkaChannel` which is
+distributed with https://github.com/knative/eventing-contrib.
```
Comment (mattmoor) [inline docs/configuration.md:325]: Remove trailing whitespace:
```suggestion
distributed with https://github.com/knative/eventing-contrib.
```

File: pkg/reconciler/knativeeventing/common/util.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"encoding/json"
```
Comment (mattmoor) [inline pkg/reconciler/knativeeventing/common/util.go:20]: Format Go code:
```suggestion
	"encoding/json"

```

File: docs/configuration.md [hunk 4]
```diff
@@ -313,3 +314,24 @@ metadata:
...
+
+```
+apiVersion: operator.knative.dev/v1alpha1
+kind: KnativeEventing
+metadata:
+  name: knative-eventing
+  namespace: knative-eventing
+spec:
+  defaultChannelTemplate:
```
Comment (matzew) [inline docs/configuration.md:334]: that would be cluster scoped?

how adding for a given set of namespaces ?   Coming to this via https://github.com/knative/docs/pull/2597/files

Comment (aliok) [reply docs/configuration.md:334]: Yeah, only cluster scope, similar to `defaultBrokerClass`.
We can also add support for namespace defaults, but in another PR.

**Comment by aliok** [comment]
/hold

Let's keep this on hold until because of the discussion at https://github.com/knative/docs/pull/2597

**Comment by aliok** [comment]
/close

As we have a generic `Config` field in CRs that allows us to set upstream configmaps, there's no need to have the field introduced in this PR.

See the discussion here: https://knative.slack.com/archives/CEPGNPHP1/p1592486833471600

**Pull request closed without merging**